From 7843f38b1604831a0099a49303e82e6c8c548dbf Mon Sep 17 00:00:00 2001
From: Jake Weinstein <jake@aospa.co>
Date: Mon, 19 Sep 2022 03:23:38 +0900
Subject: [PATCH] PropImitationHooks: Spoof Pixel properties on Google app

This exposes other widgets and functionality that
are not otherwise available, particularly
the weather widget.

Change-Id: I0b32cc5bb3e90ddf6997cd447843244a87f2fc4f
Signed-off-by: penglezos <panagiotisegl@gmail.com>
---
 .../internal/util/PropImitationHooks.java      | 18 ++++++++++++++++--
 core/res/res/values/config.xml                 |  3 +++
 core/res/res/values/symbols.xml                |  3 +++
 3 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/core/java/com/android/internal/util/PropImitationHooks.java b/core/java/com/android/internal/util/PropImitationHooks.java
index cecca63b4b83..efb55d3c1071 100644
--- a/core/java/com/android/internal/util/PropImitationHooks.java
+++ b/core/java/com/android/internal/util/PropImitationHooks.java
@@ -34,16 +34,21 @@ public class PropImitationHooks {
     private static final String sCertifiedFp =
             Resources.getSystem().getString(R.string.config_certifiedFingerprint);
 
+    private static final String sCertifiedModel =
+            Resources.getSystem().getString(R.string.config_certifiedModel);
+
     private static final String sStockFp =
             Resources.getSystem().getString(R.string.config_stockFingerprint);
 
     private static final String PACKAGE_ARCORE = "com.google.ar.core";
     private static final String PACKAGE_FINSKY = "com.android.vending";
+    private static final String PACKAGE_VELVET = "com.google.android.googlequicksearchbox";
     private static final String PACKAGE_GMS = "com.google.android.gms";
     private static final String PROCESS_GMS_UNSTABLE = PACKAGE_GMS + ".unstable";
 
     private static volatile boolean sIsGms = false;
     private static volatile boolean sIsFinsky = false;
+    private static volatile boolean sIsVelvet = false;
 
     public static void setProps(Application app) {
         final String packageName = app.getPackageName();
@@ -55,10 +60,19 @@ public static void setProps(Application app) {
 
         sIsGms = packageName.equals(PACKAGE_GMS) && processName.equals(PROCESS_GMS_UNSTABLE);
         sIsFinsky = packageName.equals(PACKAGE_FINSKY);
+        sIsVelvet = packageName.equals(PACKAGE_VELVET);
 
+        /* Set Certified Fingerprint for GMSCore or Finsky
+         * Set Certified Fingerprint and Model for Velvet
+         * Set Stock Fingerprint for ARCore
+         */
         if (!sCertifiedFp.isEmpty() && (sIsGms || sIsFinsky)) {
             dlog("Setting certified fingerprint for: " + packageName);
             setPropValue("FINGERPRINT", sCertifiedFp);
+        } else if (!sCertifiedFp.isEmpty() && (sIsVelvet)) {
+            dlog("Setting certified fingerprint and model for: " + packageName);
+            setPropValue("FINGERPRINT", sCertifiedFp);
+            setPropValue("MODEL", sCertifiedModel);
         } else if (!sStockFp.isEmpty() && packageName.equals(PACKAGE_ARCORE)) {
             dlog("Setting stock fingerprint for: " + packageName);
             setPropValue("FINGERPRINT", sStockFp);
@@ -84,8 +98,8 @@ private static boolean isCallerSafetyNet() {
 
     public static void onEngineGetCertificateChain() {
         // Check stack for SafetyNet or Play Integrity
-        if (isCallerSafetyNet() || sIsFinsky) {
-            dlog("Blocked key attestation sIsGms=" + sIsGms + " sIsFinsky=" + sIsFinsky);
+        if (isCallerSafetyNet() || sIsFinsky || sIsVelvet) {
+            dlog("Blocked key attestation sIsGms=" + sIsGms + " sIsFinsky=" + sIsFinsky + " sIsVelvet=" + sIsVelvet);
             throw new UnsupportedOperationException();
         }
     }
diff --git a/core/res/res/values/config.xml b/core/res/res/values/config.xml
index de56539cc319..a7664d7e783e 100644
--- a/core/res/res/values/config.xml
+++ b/core/res/res/values/config.xml
@@ -5936,6 +5936,9 @@
     -->
     <integer name="config_deviceStateRearDisplay">-1</integer>
 
+     <!-- Fingerprint from a certified model -->
+    <string name="config_certifiedModel" translatable="false"></string>
+
     <!-- Fingerprint from a certified device against current platform SPL level -->
     <string name="config_certifiedFingerprint" translatable="false"></string>
 
diff --git a/core/res/res/values/symbols.xml b/core/res/res/values/symbols.xml
index 8cb965de45ca..53c64e9393aa 100644
--- a/core/res/res/values/symbols.xml
+++ b/core/res/res/values/symbols.xml
@@ -4856,6 +4856,9 @@
 
   <java-symbol type="bool" name="system_server_plays_face_haptics" />
 
+  <!-- Fingerprint from a certified model -->
+  <java-symbol type="string" name="config_certifiedModel" />
+
   <!-- Fingerprint from a certified device against current platform SPL level -->
   <java-symbol type="string" name="config_certifiedFingerprint" />
 